### PR TITLE
Yolox

### PR DESCRIPTION
**Pull request by PososikTeam**

### CODE REVIEW

File: src/data/datasets/detection.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            label = self.name2label[row[self.target_cls_column]]
+            bboxes.append(bbox)
+            labels.append(label)
+        
+        image = self.augment(image=image)['image']
+
+
+        sample = {
+            'image': image,
+            'bboxes': bboxes,
```
Comment (VladVin) [inline src/data/datasets/detection.py:78]: You have to define custome `collate_fn` for bboxes and category_ids as [here](https://github.com/eora-ai/torchok/blob/main/src/data/datasets/retrieval.py#L162).
In config file it needs to be specified as [this](https://github.com/eora-ai/torchok/blob/main/src/tasks/base_task.py#L88)

File: src/models/detection/output_converter/point_generator.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import numpy as np
+import torch
+from torch.nn.modules.utils import _pair
+
+class PointGenerator:
```
Comment (VladVin) [inline src/models/detection/output_converter/point_generator.py:6]: Remove

File: src/models/detection/output_converter/sampler.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                num_expected_neg = neg_upper_bound
+        neg_inds = self.neg_sampler._sample_neg(
+            assign_result, num_expected_neg, bboxes=bboxes, **kwargs)
+        neg_inds = neg_inds.unique()
+
+        sampling_result = SamplingResult(pos_inds, neg_inds, bboxes, gt_bboxes,
+                                         assign_result, gt_flags)
+        return sampling_result
+
+class PseudoSampler(BaseSampler):
```
Comment (VladVin) [inline src/models/detection/output_converter/sampler.py:194]: Remove everything related to sampling

File: src/models/detection/output_converter/yolox.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    x1, y1, x2, y2 = bbox.split((1, 1, 1, 1), dim=-1)
+    bbox_new = [(x1 + x2) / 2, (y1 + y2) / 2, (x2 - x1), (y2 - y1)]
+    return torch.cat(bbox_new, dim=-1)
+
+@DETECTION_INFER_MODULES.register_class
+class YoloxInfer:
+    def __init__(self,
+                strides=[8, 16, 32],
+                train_cfg=None,
+                test_cfg=None,):
```
Comment (VladVin) [inline src/models/detection/output_converter/yolox.py:54]: Remove `train_cfg`, linearize `test_cfg`

File: src/models/losses/iou.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+
+def fp16_clamp(x, min=None, max=None):
+    if not x.is_cuda and x.dtype == torch.float16:
+        # clamp for cpu float16, tensor fp16 has no clamp implementation
+        return x.float().clamp(min, max).half()
+
+    return x.clamp(min, max)
+
+def bbox_overlaps(bboxes1, bboxes2, mode='iou', is_aligned=False, eps=1e-6):
```
Comment (VladVin) [inline src/models/losses/iou.py:23]: Keep one version of this code

File: src/models/detection/output_converter/yolox.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (VladVin) [inline src/models/detection/output_converter/yolox.py:1]: Rename to `src/models/detection/infer/yolox.py`

File: src/tasks/detection_task.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+class DetectionTaskParams(BaseModel):
+    checkpoint: str = None
+    input_size: list  # [num_channels, height, width]
+    freeze_backbone: bool = False
+    backbone_name: str
+    backbone_params: dict = {}
+    neck_name: str
+    neck_params: dict = {}
+    head_name: str
+    head_params: dict = {}
```
Comment (VladVin) [inline src/tasks/detection_task.py:19]: Add `infer_params`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                self.transform,
+                bbox_params=module_transforms.BboxParams(
+                    format='pascal_voc',
+                    label_fields=['category_ids'],
+                    min_area=min_area,
+                    min_visibility=min_visibility
+                    )
+            )
+
+        self.csv[target_column] = self.csv[target_column].apply(eval)
```
Comment (VladVin) [inline src/data/datasets/detection.py:63]: `eval` variable not defined

File: src/data/datasets/detection.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            'image': image,
+            'bboxes': bboxes,
+            'category_ids': labels
+            }
+
+        if self.augment is not None:
+            sample = self.augment(**sample)
+
+        sample = self.transform(**sample)
+        sample['bbox_count'] = bbox_count
```
Comment (VladVin) [inline src/data/datasets/detection.py:103]: Since Albumentations can remove some bbox, the real number of bboxes can change. You should calc count after augmentations are applied

File: src/data/datasets/detection.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        
+        return sample
+
+    @staticmethod
+    def collate_fn(batch: dict) -> dict:
+        """
+        Add empty bbox and label into batch with different size of bboxes
```
Comment (VladVin) [inline src/data/datasets/detection.py:110]: Pad bboxes and labels tensors with empty data to form a fix shaped output tensors. Size of the corresponding dimension is equal to the maximum number of bboxes in the given batch

File: src/data/datasets/detection.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        empty bbox = [0, 0, 0, 0]
+        empty label = -1
+        """
+        # get sequence lengths
```
Comment (VladVin) [inline src/data/datasets/detection.py:114]: `get maximum sequence length`

File: src/data/datasets/detection.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        max_length = 0
+        for t in batch:
+            max_length = max(max_length, t['bbox_count'])
+       
+        empty_box = [0]*4
+        for t in batch:
+            bbox_count = t['bbox_count']
```
Comment (VladVin) [inline src/data/datasets/detection.py:121]: Make it simple and faster:
```
bboxes = torch.zeros(max_length, 4, dtype=torch.float16)
bboxes[:t['bbox_count']] = t['target_bboxes']
labels = torch.full(max_length, value=-1, dtype=torch.long)
labels[:t['bbox_count']] = t['target_labels']
```

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py (hunk 8 above)
Comment (PososikTeam) [reply src/data/datasets/detection.py:63]: eval its defined pandas variable. it convert '[1, 2, 3, 4]' : str -> [1, 2, 3, 4] : List[int]

Comment (PososikTeam) [reply src/data/datasets/detection.py:63]: example https://www.kaggle.com/debarshichanda/w-b-tables-great-barrier-reef-eda

File: notebooks/CheckTransforms.ipynb [hunk 13]
```diff
@@ -0,0 +1,670 @@
+{
```
Comment (VladVin) [inline notebooks/CheckTransforms.ipynb:1]: Let's create a separate `CheckTransformsDetection` and put it to `examples/notebooks`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py (hunk 12 above)
Comment (PososikTeam) [reply src/data/datasets/detection.py]: about float16, we have x_min, y_min, x_max, y_max format, we dont care about long and change into float16?

File: src/data/datasets/detection.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        self.csv[target_column] = self.csv[target_column].apply(eval)
+
+    def __getitem__(self, idx: int):
+        sample = self.get_raw(idx // self.expand_rate)
+        sample['image'] = sample['image'].type(torch.__dict__[self.input_dtype])
+        
+        output = {
+            'input': sample['image'],
+            'target_bboxes': torch.tensor(sample['bboxes']).type(torch.__dict__[self.target_dtype]),
+            'target_labels': torch.tensor(sample['category_ids']).type(torch.__dict__[self.target_dtype]),
```
Comment (VladVin) [inline src/data/datasets/detection.py:72]: Can you name it `target_classes`, please? Because `labels` is a more general word

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/detection/heads/yolox.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+            blocks. Default: False
+        dcn_on_last_conv (bool): If true, use dcn in the last layer of
+            towers. Default: False.
+        conv_bias (bool | str): If specified as `auto`, it will be decided by
+            the norm_cfg. Bias of conv will be set as True if `norm_cfg` is
+            None, otherwise False. Default: "auto".
+        conv_cfg (dict): Config dict for convolution layer. Default: None.
+        norm_cfg (dict): Config dict for normalization layer. Default: None.
+        act_cfg (dict): Config dict for activation layer. Default: None.
+        loss_cls (dict): Config of classification loss.
```
Comment (VladVin) [inline src/models/detection/heads/yolox.py:58]: Remove unused parameters

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/losses/cross_entropy.py [hunk 16]
```diff
@@ -43,6 +43,8 @@ def __init__(self, weight=None, reduction='mean', pos_weight=None, ignore_all_ze
         self.ignore_all_zeros = ignore_all_zeros
 
     def forward(self, input, target):
+        if input.numel() == 0:
```
Comment (VladVin) [inline src/models/losses/cross_entropy.py:46]: Don't customize loss for this problem, return `torch.tensor(0.)` from task

File: src/models/losses/iou.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+            target (torch.Tensor): Corresponding gt bboxes, shape (n, 4).
+            linear (bool, optional): If True, use linear scale of loss instead of
+                log scale. Default: False.
+            mode (str): Loss scaling mode, including "linear", "square", and "log".
+                Default: 'log'
+            eps (float): Eps to avoid log(0).
+        Return:
+            torch.Tensor: Loss tensor.
+        """
+        if input.numel() == 0:
```
Comment (VladVin) [inline src/models/losses/iou.py:149]: Remove

File: src/tasks/detection_task.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
+from typing import DefaultDict
+import torch
+from pydantic import BaseModel
+
+from src.constructor import create_backbone, create_scheduler, create_optimizer
+from src.constructor.config_structure import TrainConfigParams
+from src.registry import TASKS, DETECTION_NECKS, \
+    DETECTION_HEADS, DETECTION_HAT
```
Comment (VladVin) [inline src/tasks/detection_task.py:8]: `DETECTION_HATS`

File: src/tasks/detection_task.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        return output
+
+    def training_step(self, batch, batch_idx):
+        output = self.forward_with_gt(batch)
+        loss = self.criterion(**output)
+        input_data = batch['input']
+        gt_bboxes = batch['target_bboxes']
+        gt_labels = batch['target_labels']
+        prediction = self.forward_valid(input_data)
```
Comment (VladVin) [inline src/tasks/detection_task.py:125]: Pass original bboxes and everithing

File: src/tasks/detection_task.py [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+            'cls_targets': loss_cls_targets,
+            }
+        return output
+
+    def training_step(self, batch, batch_idx):
+        output = self.forward_with_gt(batch)
+        loss = self.criterion(**output)
+        input_data = batch['input']
+        gt_bboxes = batch['target_bboxes']
+        gt_labels = batch['target_labels']
```
Comment (VladVin) [inline src/tasks/detection_task.py:124]: Rename to `target_classes`

File: src/tasks/detection_task.py [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+        prediction = self.forward_valid(input_data)
+        target = [[gt_bboxes[i], gt_labels[i]] for i in range(gt_bboxes.shape[0])]
```
Comment (VladVin) [inline src/tasks/detection_task.py:126]: `target = list(zip(gt_bboxes, gt_labels))`

File: src/tasks/detection_task.py [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+        valid_output = {
+            'target': target,
+            'prediction': prediction
+        }
+        self.metric_manager.update('valid', **valid_output)
+        return loss
+
+    # def test_step(self, batch, batch_idx):
+    #     output = self.forward_with_gt(batch)
```
Comment (VladVin) [inline src/tasks/detection_task.py:150]: Call `validation_step`

File: src/data/datasets/detection.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        # get maximum sequence length
+        max_length = 0
+        for t in batch:
+            max_length = max(max_length, t['bbox_count'])
+
+        if max_length != 0:
+            for t in batch:
+                bboxes = torch.zeros(max_length, 4, dtype=torch.float16)
+                labels = torch.full((max_length,), -1, dtype=torch.long)
```
Comment (VladVin) [inline src/data/datasets/detection.py:121]: `classes`

File: src/models/losses/iou.py [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+    # calculate gious
+    enclose_wh = (enclosed_rb - enclosed_lt).clamp(min=0)
+    enclose_area = enclose_wh[..., 0] * enclose_wh[..., 1]
+    enclose_area = torch.max(enclose_area, eps)
+    gious = ious - (enclose_area - union) / enclose_area
+    return gious
+
+
+@LOSSES.register_class
+class DetectionIoU(nn.Module):
```
Comment (VladVin) [inline src/models/losses/iou.py:125]: `DetectionIoULoss`

File: src/tasks/detection_task.py [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+        )
+        
+        head_class = DETECTION_HEADS.get(self.params.head_name)
+        self.head = head_class(
+                            num_classes=self.params.head_params['num_classes'], 
+                            in_channels=128, 
+                            feat_channels=128
+        )
+
+        infer_class = DETECTION_HAT.get(self.params.hat_name)
```
Comment (VladVin) [inline src/tasks/detection_task.py:51]: `hat_class`

File: src/tasks/detection_task.py [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+        self.infer_module = infer_class(num_classes=self.params.head_params['num_classes'])
+        self.num_classes = self.params.head_params['num_classes']
+
+        self._set_pretrained_weights()
+
+    def _set_pretrained_weights(self, url = 'https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_s_8x8_300e_coco/yolox_s_8x8_300e_coco_20211121_095711-4592a793.pth'):
```
Comment (VladVin) [inline src/tasks/detection_task.py:57]: Implement separate loading of state dict for backbone and neck

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/tasks/detection_task.py (hunk 25 above)
Comment (PososikTeam) [reply src/tasks/detection_task.py:51]: its already named backbone _name, neck_name and hat_name, if change hat_name to hat class we need to change backbone_name to backbone_class

File: src/data/datasets/detection.py [hunk 27]
```diff
@@ -0,0 +1,132 @@
+import cv2
```
Comment (VladVin) [inline src/data/datasets/detection.py:1]: Remove unused imports

File: src/data/datasets/detection.py [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+
+        self.csv[target_column] = self.csv[target_column].apply(eval)
+
+    def __getitem__(self, idx: int):
+        sample = self.get_raw(idx // self.expand_rate)
+        sample['image'] = sample['image'].type(torch.__dict__[self.input_dtype])
+        
+        output = {
+            'input': sample['image'],
+            'target_bboxes': torch.tensor(sample['bboxes']).type(torch.float16),
```
Comment (VladVin) [inline src/data/datasets/detection.py:71]: Type: `torch.__dict__[self.target_dtype]`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+                    format='pascal_voc',
+                    label_fields=['category_ids'],
+                    min_area=min_area,
+                    min_visibility=min_visibility
+                    )
+            )
+
+        self.csv[target_column] = self.csv[target_column].apply(eval)
+
+    def __getitem__(self, idx: int):
```
Comment (VladVin) [inline src/data/datasets/detection.py:65]: Put documentation what's returned from this function

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+        sample = self.get_raw(idx // self.expand_rate)
+        sample['image'] = sample['image'].type(torch.__dict__[self.input_dtype])
+        
+        output = {
+            'input': sample['image'],
+            'target_bboxes': torch.tensor(sample['bboxes']).type(torch.float16),
+            'target_classes': torch.tensor(sample['category_ids']).type(torch.__dict__[self.target_dtype]),
```
Comment (VladVin) [inline src/data/datasets/detection.py:72]: `category_ids` -> `classes`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/data/datasets/detection.py (hunk 30 above)
Comment (VladVin) [reply src/data/datasets/detection.py:72]: Put type `torch.long`

**Review by VladVin — CHANGES_REQUESTED** [review]
